proposals/opt-in.md

```diff
@@ -0,0 +1,289 @@
+    * We've explored the possibility of using a “string tag” argument instead of custom marker annotations (e.g. `@RequiresOptIn(“kotlin.ExperimentalTypeInference”) ...`) but discarded it because it's not as clean, involves more typing and reading, requires not to make typos, and complicates the implementation, especially in the IDE.
+* There's a number of ways to express the opt-in to use an API, but a source-retained annotation (“local” opt-in) and a command line argument (“global” opt-in) would seem the most natural choices.
+
+## Proposal
+
+We propose to add the following declarations to the standard Kotlin library:
+
+```kotlin
+package kotlin
+
+@Target(ANNOTATION_CLASS)
+@Retention(BINARY)
+annotation class RequiresOptIn(
+    val message: String = "",
+    val level: Level = Level.ERROR
+) {
+    enum class Level { WARNING, ERROR }
+}
+
+@Target(CLASS, PROPERTY, LOCAL_VARIABLE, VALUE_PARAMETER, CONSTRUCTOR, FUNCTION,
+        PROPERTY_GETTER, PROPERTY_SETTER, EXPRESSION, FILE, TYPEALIAS)
+@Retention(SOURCE)
+annotation class OptIn(
+    vararg val markerClass: KClass<out Annotation>
+)
+```
+
+The `RequiresOptIn` annotation is applied to an annotation class, and it makes that class an **opt-in requirement marker**. There are two sets of use cases where markers are used:
+
+1. If a declaration is annotated with the marker, it requires opt-in to that marker and can use other API *with that same marker* in its body.
+2. If a declaration or an expression is annotated with `@OptIn(Marker::class)`, it can use other declarations that use the selected marker, but it does not require opt-in itself (its clients will not *have* to opt in).
+
+The first option of usage of opt-in requirement markers is called a **propagating opt-in** (the annotation effectively causes propagation of the requirement), and the second — a **non-propagating opt-in**. The user is free to choose whichever option is preferable in each scenario.
+
+Example:
+
+```kotlin
+// Library code:
+
+@RequiresOptIn
+annotation class ShinyNewAPI
+
+// Class Foo requires opt-in with marker ShinyNewAPI
+@ShinyNewAPI
+class Foo { ... }
+
+// Function bar requires opt-in with marker ShinyNewAPI
+@ShinyNewAPI
+fun Foo.bar() = ...
+
+// Usage:
+
+// Function useFooAndBar uses API marked with ShinyNewAPI and thus
+// is required to opt in to that marker. Here, we choose a propagating opt-in,
+// because the API is used in the function signature (parameter foo's type).
+// (But this is not enforced by the compiler in any way, we could've chosen
+// a non-propagating opt-in as well.)
+@ShinyNewAPI
+fun useFooAndBar(foo: Foo) {
```

Comment:
This is an outdated example and comment that was made obsolete with the recent changes. I suggest expanding an example section:

1. More this `useFooAndBar` to the end of the examples, mark it with `@OptIn` and explain that, even though it is an `@OptIn`, every usage will still get a warning or error, because it references an opt-in type.

2. Change `@ShinyNewAPI`-marked example here, so that it is similar to the `doSomething` below (does not have any parameters) but its author chose to mark it with `@ShinyNewAPI`. Name it something like `shinyNewFeature` and explain that this function is also, conceptually, a part of `ShinyNewAPI` and so it is marked as such. For contrast, `doSomething` example's comment can explain that this is an old function that is was always a part of the library and the fact that it is now using shiny new API is just its implementation detail.

Reply:
I'd just delete the first example, because contagiousness is described at the end of the document separately.

Reply:
The second example was split into two different as you wish